### PR TITLE
Hardcoded features for HotA-like starting conditions

### LINKED ISSUE #1520: HotA's system of starting troops.
**Describe your proposal**
In HotA heroes with a starting army **always** have all troops (and a war machine), without the possibility of absence of any of them. 

OH3 mechanic is:

The probability of heroes starting (or appearing in Tavern) with only 1 stack is 9%, with 2 its 69%,
with 3 its 22%.
Level 1 unit specialists will only start with level 1 units.
Level 2 unit specialists will only start with level 2 units.
Level 3 unit specialists may start with level 1 and level 3 units.
Heroes having the possibility to start with a War Machine will not start with level 2 units.

**Additional context**
I think it will be cool to have an opportunity to switch mechanics in defaultMods.json

### PR DESCRIPTION
**Pull request by dydzio0614**
I am not sure about war machines part, but for other stuff it fixes #1520 and it's improvement overall for some people who want to play with this feature (got some questions about it).

### CODE REVIEW

**Comment by IvanSavenko** [comment]
Perhaps better to define chances instead?
```
// In H3 hero has 10% chance to start with 1 stack, 70% for 2 stacks, and 20% for 3 stacks
HERO_STARTING_ARMY_STACKS_COUNT = [ 10, 70, 20 ]
// In HotA, hero always starts with 3 stacks
HERO_STARTING_ARMY_STACKS_COUNT = [ 0, 0, 100 ]

// In H3, town with default building set will always have 1st dwelling and 50% chance for second
DEFAULT_BUILDING_SET_DWELLING_CHANCES = [ 100, 50 ]
// In HotA, towns with default building set will always have 1st and 2nd dwelling
DEFAULT_BUILDING_SET_DWELLING_CHANCES = [ 100, 50 ]
```

**Comment by dydzio0614** [comment]
Planned to do it as you say now, but after all I think it's way less intuitive to use and restore default value

**Comment by IvanSavenko** [comment]
I view this mostly from point of removing hardcoded values from engine. There is no point in hardcoding values that can be loaded from config.

As for ease of editing - file in vcmi mod is not supposed to be edited by players, and modders who replace it can always check version from vcmi mod.

**Comment by dydzio0614** [comment]
This file isn't in VCMI mod, it is in config folder. It is pretty mod-independent as it changes game rules that are not useful to be parts of mod features.

**Comment by IvanSavenko** [comment]
Still, our config files are not supposed to be edited by players. So "easier to restore original value" is a bad argument IMO.

And in long-term I think this file should only be edited by mods. "Game balancing" is a mod feature after all.

**Comment by dydzio0614** [comment]
Long term is long term, but in meantime players want some way to be able to access this feature

**Comment by IvanSavenko** [comment]
Make a small mod that has this field correctly set?
I *really* dislike various "temporary workaround". Why bother with workarounds when it can be done properly?

**Comment by misiokles** [comment]
Creating mini-mods that can alter game configs is rather simple way. Some mod eneble negative luck in such way.

**Comment by dydzio0614** [comment]
OK, though I still do not see it as fully proper way - proper way would allow heroes to have all 7 stacks filled by default etc.

**Comment by IvanSavenko** [comment]
> OK, though I still do not see it as fully proper way - proper way would allow heroes to have all 7 stacks filled by default without hardcoding exactly three array values etc.

Then don't hardcode? Use std::vector instead of std::array and load every existing value from file.
So if heroes get more than 3 stacks game will still load correct values

File: lib/CModHandler.h [hunk 1]
```diff
@@ -382,6 +384,8 @@ class DLL_LINKAGE CModHandler
 			h & ATTACK_POINTS_DMG_MULTIPLIER_CAP;
 			h & DEFENSE_POINT_DMG_MULTIPLIER;
 			h & DEFENSE_POINTS_DMG_MULTIPLIER_CAP;
+			h & NEW_HERO_ALWAYS_3_CREATURE_STACKS;
+			h & DEFAULT_TOWN_ALWAYS_2_DWELLINGS;
```
Comment (Nordsoft91) [inline lib/CModHandler.h:388]: please, bump serialisation version and also you can provide backward compatibility easy here:
```suggestion
		if(version >= xxx)
		{
			h & NEW_HERO_ALWAYS_3_CREATURE_STACKS;
			h & DEFAULT_TOWN_ALWAYS_2_DWELLINGS;
		}
```

**Review by Nordsoft91 — CHANGES_REQUESTED** [review]

File: lib/CModHandler.cpp [hunk 2]
```diff
@@ -767,6 +767,18 @@ void CModHandler::loadConfigFromFile (std::string name)
 	settings.DEFENSE_POINTS_DMG_MULTIPLIER_CAP = hardcodedFeatures["DEFENSE_POINTS_DMG_MULTIPLIER_CAP"].Float();
 	logMod->debug("\tDEFENSE_POINTS_DMG_MULTIPLIER_CAP\t%f", settings.DEFENSE_POINTS_DMG_MULTIPLIER_CAP);
 
+	JsonVector stacksChancesVector = hardcodedFeatures["HERO_STARTING_ARMY_STACKS_COUNT_CHANCES"].Vector();
```
Comment (IvanSavenko) [inline lib/CModHandler.cpp:770]: use json.convertTo<std::vector<int>>();

File: lib/mapObjects/CGHeroInstance.cpp [hunk 3]
```diff
@@ -348,16 +348,25 @@ void CGHeroInstance::initArmy(CRandomGenerator & rand, IArmyDescriptor * dst)
 	if(!dst)
 		dst = this;
 
-	int howManyStacks = 0; //how many stacks will hero receives <1 - 3>
-	int pom = rand.nextInt(99);
+	int howManyStacks = 7; //how many stacks will hero receives <1 - 7>
```
Comment (IvanSavenko) [inline lib/mapObjects/CGHeroInstance.cpp:351]: Move to actual initialization
```suggestion
```

File: lib/mapObjects/CGHeroInstance.cpp [hunk 4]
```diff
@@ -348,16 +348,25 @@ void CGHeroInstance::initArmy(CRandomGenerator & rand, IArmyDescriptor * dst)
...
+	if(allowZeroStacksArmy)
+		stacksCountChances.pop_back();
+
+	int stacksCountInitRandomNumber = rand.nextInt(1, 100);
+
+	auto stacksCountElementIndex = vstd::find_pos_if(stacksCountChances, [stacksCountInitRandomNumber](int element){ return stacksCountInitRandomNumber < element; });
+	if(stacksCountElementIndex == -1)
+		stacksCountElementIndex = stacksCountChances.size();
+
+	howManyStacks = stacksCountElementIndex;
```
Comment (IvanSavenko) [inline lib/mapObjects/CGHeroInstance.cpp:367]:
```suggestion
	int howManyStacks = stacksCountElementIndex;
```

File: lib/mapObjects/CGHeroInstance.cpp [hunk 5]
```diff
@@ -348,16 +348,25 @@ void CGHeroInstance::initArmy(CRandomGenerator & rand, IArmyDescriptor * dst)
...
-	if(pom < 9)
-		howManyStacks = 1;
-	else if(pom < 79)
-		howManyStacks = 2;
-	else
-		howManyStacks = 3;
+	std::vector<int32_t> stacksCountChances = VLC->modh->settings.HERO_STARTING_ARMY_STACKS_COUNT_CHANCES;
+
+	const int zeroStacksAllowingValue = -1;
+	bool allowZeroStacksArmy = !stacksCountChances.empty() && stacksCountChances.back() == zeroStacksAllowingValue;
```
Comment (IvanSavenko) [inline lib/mapObjects/CGHeroInstance.cpp:357]: Not sure if this check is needed - I expect vcmi to break on empty armies anyway.

**Review by IvanSavenko — APPROVED** [review]

File: lib/mapObjects/CGHeroInstance.cpp (hunk 5 above)
Comment (dydzio0614) [reply lib/mapObjects/CGHeroInstance.cpp:357]: empty armies work in general so this feature should stay

**Comment by IvanSavenko** [comment]
@Nordsoft91 
OK to merge?

**Review by Nordsoft91 — APPROVED** [review]

**Pull request merged**